src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -1205,29 +1205,24 @@ private void Register()
             Debug.Assert(_nonBlockingSet);
             lock (_registerLock)
             {
-                if (!_registered)
+                if (_asyncEngine == null)
                 {
-                    Debug.Assert(!_asyncEngineToken.WasAllocated);
-                    var token = new SocketAsyncEngine.Token(this);
+                    bool addedRef = false;
+                    try
+                    {
+                        _socket.DangerousAddRef(ref addedRef);
```

Comment:
DangerousAddRef will throw ODE in the addRef == false case.

Reply:
> DangerousAddRef will throw ODE in the addRef == false case.

Weird API. What's the point of the `out bool addRef` parameter then?

------------------------------------

What about the CER concern? The [docs](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.safehandle.dangerousaddref) state:

> To avoid leaking handle resources, always call this method inside a constrained execution region (CER), where a thread abort cannot interrupt processing.

Although I've seen other points BCL code where the recommendation was ignored, I'm confused about it.

Reply:
@antonfirsov I'm not an expert. This is my understanding. Thread aborting is not implemented on .NET Core Linux (maybe even on Windows). I think it's more important that `DangerousRelease` is in a CER than `DangerousAddRef`.

Reply:
Thread aborts don't exist in .NET Core (other than via some debugger interactions) and CERs are nops.